nomad/deploymentwatcher/batcher.go

```diff
@@ -0,0 +1,141 @@
+package deploymentwatcher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// EvalBatcher is used to batch the creation of evaluations
+type EvalBatcher struct {
+	// batch is the batching duration
+	batch time.Duration
+
+	// raft is used to actually commit the evaluations
+	raft DeploymentRaftEndpoints
+
+	// future to be returned to callers
+	f *EvalFuture
+
+	// inCh is used to pass evaluations to the daemon process
+	inCh chan *structs.Evaluation
+
+	// ctx is used to exit the daemon batcher
+	ctx context.Context
+
+	l sync.Mutex
+}
+
+// NewEvalBatcher returns an EvalBatcher that uses the passed raft endpoints to
+// create the evaluations and exits the batcher when the passed exit channel is
+// closed.
+func NewEvalBatcher(batchDuration time.Duration, raft DeploymentRaftEndpoints, ctx context.Context) *EvalBatcher {
+	b := &EvalBatcher{
+		batch: batchDuration,
+		raft:  raft,
+		ctx:   ctx,
+		inCh:  make(chan *structs.Evaluation, 10),
+	}
+
+	go b.batcher()
+	return b
+}
+
+// CreateEval batches the creation of the evaluation and returns a future that
+// tracks the evaluations creation.
+func (b *EvalBatcher) CreateEval(e *structs.Evaluation) *EvalFuture {
+	b.l.Lock()
+	if b.f == nil {
+		b.f = NewEvalFuture()
+	}
+	b.l.Unlock()
+
+	b.inCh <- e
+	return b.f
+}
+
+// batcher is the long lived batcher goroutine
+func (b *EvalBatcher) batcher() {
+	timer := time.NewTimer(b.batch)
+	evals := make(map[string]*structs.Evaluation)
+	for {
+		select {
+		case <-b.ctx.Done():
+			timer.Stop()
+			return
+		case e := <-b.inCh:
+			if len(evals) == 0 {
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(b.batch)
+			}
+
+			evals[e.DeploymentID] = e
+		case <-timer.C:
+			if len(evals) == 0 {
+				// Reset the timer
+				timer.Reset(b.batch)
+				continue
+			}
+
+			// Capture the future
+			b.l.Lock()
+			f := b.f
+			b.f = nil
+			b.l.Unlock()
+
+			// Shouldn't be possible but protect ourselves
+			if f == nil {
+				// Reset the timer
+				timer.Reset(b.batch)
+				continue
+			}
+
+			// Capture the evals
+			all := make([]*structs.Evaluation, 0, len(evals))
+			for _, e := range evals {
+				all = append(all, e)
+			}
+
+			// Upsert the evals
+			f.Set(b.raft.UpsertEvals(all))
```

Comment:
We would lose ordering guarantees and could have an unbounded number of upserts in flight. I'm not sure if those are acceptable "losses" for this block of code.

`inCh` being buffered at lets another batch be built concurrently while maintaining ordering and a bound on the eval buffer.